### DISCUSSION TITLE
Check for pending/outstanding changes

### DISCUSSION THREAD

**Discussion by sondrelg** [sqlalchemy/alembic#1072] [Usage Questions] [answered]
As a part of our CI/CD pipelines for Django projects, we tend to include a check for any model changes that haven't been reflected in a migration file, using:

```python
python manage.py makemigrations --check --dry-run
```

This exits with exit code 1 if the Django migration framework detects new changes, and fails our pipeline. This is great, since we generally want all model changes to be declared in migration files before deploying our application. 

Does something like this exist for `alembic`? I wasn't able to find anything in the CLI or previous discussions/issues 🙂

**Comment by zzzeek** [comment, marked as answer]
yes, there's a recipe for this case [here](https://alembic.sqlalchemy.org/en/latest/api/autogenerate.html#alembic.autogenerate.compare_metadata) (that should really be in the cookbook) and there are projects on pypi that do this also such as  https://github.com/4Catalyzer/alembic-autogen-check  https://pypi.org/project/sqlalchemy-diff/

**Reply by sondrelg** [reply]
Thanks @zzzeek 👏  Ended up writing this, in case it's helpful for anyone else. Seems to do the trick

```python
import os
import pprint
import sys

from alembic.autogenerate import compare_metadata
from alembic.runtime.migration import MigrationContext
from sqlalchemy import create_engine

# Import metadata here

engine = create_engine(str(os.environ['DATABASE_URL'].replace('+asyncpg', '')))
metadata = BaseMeta.metadata
migration_context = MigrationContext.configure(engine.connect())

if diff := compare_metadata(migration_context, BaseMeta.metadata):
    pprint.pprint(diff, indent=2, width=10)
    print(
        'There are changes in the database models not reflect in migration files. '
        'Create migration files for these by running `alembic revision --autogenerate '
        '-m "your message"´'
    )
    sys.exit(1)
else:
    print('Migrations are up to date')
    sys.exit(0)
```
